### PR TITLE
Stricter handling of CLI args

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This implements stricter handling of CLI args by:

* Raising a NoMethodError if you call an arg method without the switch/flag being defined.
  - This unveiled quite a few bugs of broken arg checks.
* Changing the switch semantics:
  - Regular switches are now strictly boolean true/false, rather than being either true or nil.
  - "Binary" (more like ternary) switch options (`--[no-]switch`) remain three-state: true, false or nil.

This required a number of changes elsewhere to do this. I suspect there will be more I've missed, and potentially some in other Homebrew repositories which the CI may flag.

Feel free to ask about any part of the changes and I can explain it better.

### CODE REVIEW

**Comment by Bo98** [comment]
```
cli/args.rb:137:7: C: When using method_missing, define respond_to_missing?.
      def method_missing(mid, *args) ...
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

->

```rb
def respond_to_missing?(*)
  # rubocop wants us to define this, but the super implementation is sufficient.
  super
end
```

```
cli/args.rb:137:7: W: [Correctable] Useless method definition detected.
      def respond_to_missing?(*) ...
      ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

??

File: Library/Homebrew/cli/args.rb [hunk 1]
```diff
@@ -129,12 +127,22 @@ def cli_args
...
           :head
         else
           default
         end
       end
+
+      def method_missing(mid, *args)
+        ret = super
+
+        if frozen? && mid =~ /.*(?!=\z)/ && args.length == 0 && !@table.key?(mid)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:140]: Could this get a comment, not obvious what's going on here?

File: Library/Homebrew/cli/args.rb [hunk 2]
```diff
@@ -129,12 +127,22 @@ def cli_args
...
+          raise NoMethodError, "CLI arg for `#{mid}' is not declared for this command"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:141]:
```suggestion
          raise NoMethodError, "CLI arg for `#{mid}` is not declared for this command"
```

File: Library/Homebrew/cli/args.rb [hunk 3]
```diff
@@ -129,12 +127,22 @@ def cli_args
...
-        if HEAD?
+        if @table[:HEAD?]
           :head
         else
           default
         end
       end
+
+      def method_missing(mid, *args)
+        ret = super
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:138]:
```suggestion
        return_value = super
```

File: Library/Homebrew/cli/args.rb [hunk 4]
```diff
@@ -129,12 +127,22 @@ def cli_args
...
       def spec(default = :stable)
-        if HEAD?
+        if @table[:HEAD?]
           :head
         else
           default
         end
       end
+
+      def method_missing(mid, *args)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:137]:
```suggestion
      def method_missing(method_name, *args)
```

File: Library/Homebrew/cli/args.rbi [hunk 5]
```diff
@@ -3,145 +3,145 @@
 module Homebrew
   module CLI
     class Args < OpenStruct
-      sig { returns(T.nilable(T::Boolean)) }
+      sig { returns(T::Boolean) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rbi:6]: Nice 😎

**Review by MikeMcQuaid — APPROVED** [review]
Nice work here.

**Review by Rylan12 — APPROVED** [review]
Looks good to me

**Comment by Bo98** [comment]
I'll probably merge this after the next tag since it's quite likely there's something that breaks from this (see: all prior args changes).

**Comment by MikeMcQuaid** [comment]
> I'll probably merge this after the next tag since it's quite likely there's something that breaks from this (see: all prior args changes).

Tagged 3.0.8 now.

**Comment by carlocab** [comment]
This broke `brew list`:
```
❯ brew list
[snip]
Error: CLI arg for `appdir` is not declared for this command
Please report this issue:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/cli/args.rb:147:in `method_missing'
/usr/local/Homebrew/Library/Homebrew/cask/config.rb:45:in `from_args'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/list.rb:45:in `list_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/list.rb:158:in `list_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/list.rb:120:in `list'
/usr/local/Homebrew/Library/Homebrew/brew.rb:122:in `<main>'
```

From `git bisect`:
```
8e98ce69f39cf014f63195ed880d3f445c905757 is the first bad commit
commit 8e98ce69f39cf014f63195ed880d3f445c905757
Author: Bo Anderson <mail@boanderson.me>
Date:   Thu Mar 18 14:46:48 2021 +0000

    Stricter handling of CLI args
```

**Comment by Bo98** [comment]
I already have https://github.com/Homebrew/brew/pull/10905 open to fix that.

**Pull request merged**